dal/db_race_test.go

```diff
@@ -1,95 +1,97 @@
 	assert.NilError(t, err)
 	err = db.InitDB()
 	assert.NilError(t, err)
 
-	const batch = 1000
-	const workers = 5
-	done := make(chan bool, workers)
-
+	// Test 1: Parallel inserts
+	var wg1 sync.WaitGroup
 	for i := uint64(0); i < workers; i++ {
-		// TODO: use WorkGroup
-		go insertManySignReq(t, done, db, i*batch, (i+1)*batch)
-	}
-	for i := 0; i < workers; i++ {
-		<-done
+		wg1.Add(1)
+		go func(i uint64) {
+			defer wg1.Done()
+			insertManySignReq(t, db, i*batch, (i+1)*batch)
+		}(i)
 	}
+	wg1.Wait()
 
 	var srID uint64 = 1
 	sr, err := db.GetIkaSignRequestByID(srID)
 	assert.NilError(t, err)
 	assert.Assert(t, sr != nil)
-	assert.Equal(t, int64(123), sr.Timestamp)
+	assert.Equal(t, testTimestamp, sr.Timestamp)
 
-	row := db.conn.QueryRow(`
-SELECT COUNT(*) FROM ika_sign_requests`)
+	row := db.conn.QueryRow(`SELECT COUNT(*) FROM ika_sign_requests`)
 	var count int
 	err = row.Scan(&count)
 	assert.NilError(t, err)
 	assert.Equal(t, batch*workers, count)
 
-	// test 2
-	// try parallel update
-
+	// Test 2: Parallel updates
+	var wg2 sync.WaitGroup
 	for i := uint64(0); i < workers; i++ {
-		go loopIncrementIkaSRTimestamp(t, done, db, batch, srID)
-	}
-	for i := 0; i < workers; i++ {
-		<-done
+		wg2.Add(1)
+		go func(i uint64) {
+			defer wg2.Done()
+			loopIncrementIkaSRTimestamp(t, db, batch, srID)
+		}(i)
 	}
+	wg2.Wait()
 
 	sr, err = db.GetIkaSignRequestByID(srID)
 	assert.NilError(t, err)
 	assert.Assert(t, sr != nil)
-	assert.Equal(t, int64(123+workers*batch), sr.Timestamp)
+	assert.Equal(t, int64(testTimestamp+workers*batch), sr.Timestamp)
 
 }
-func insertManySignReq(t *testing.T, done chan<- bool, db DB, idFrom, idTo uint64) {
+
+func insertManySignReq(t *testing.T, db DB, idFrom, idTo uint64) {
 	for i := idFrom; i < idTo; i++ {
 		sr := IkaSignRequest{
 			ID:        i,
 			Payload:   []byte{},
 			DWalletID: "",
 			UserSig:   "",
 			FinalSig:  nil,
-			Timestamp: 123,
+			Timestamp: testTimestamp,
 		}
 		err := db.InsertIkaSignRequest(sr)
 		assert.NilError(t, err)
+		t.Logf("Worker %d finished inserting %d sign requests (from ID %d to %d)", i, idTo-idFrom, idFrom, idTo)
 	}
-
-	t.Logf("finished from %d, to: %d", idFrom, idTo)
-	done <- true
 }
 
-func loopIncrementIkaSRTimestamp(t *testing.T, done chan<- bool, db DB, n int, srID uint64) {
+func loopIncrementIkaSRTimestamp(t *testing.T, db DB, n int, srID uint64) {
 	for i := 0; i < n; i++ {
-		db.incrementIkaSRTimestamp(t, srID)
+		assert.NilError(t, db.incrementIkaSRTimestamp(srID))
 	}
-	done <- true
 }
 
-func (db DB) incrementIkaSRTimestamp(t *testing.T, srID uint64) {
+func (db DB) incrementIkaSRTimestamp(srID uint64) error {
 	db.mutex.Lock()
 	defer db.mutex.Unlock()
 
-	row := db.conn.QueryRow(`
-SELECT timestamp FROM ika_sign_requests WHERE id = ?`, srID)
+	row := db.conn.QueryRow(`SELECT timestamp FROM ika_sign_requests WHERE id =?`, srID)
 	var tm int64
-	assert.NilError(t, row.Scan(&tm))
+	if err := row.Scan(&tm); err != nil {
+		return err
+	}
```

Comment:
why returning error? This function is defined in tests, so it's not exported.

Reply:
I return error here to handle the error in the calling function `loopIncrementTimestamp`

Reply:
but if we pass `testing.T` then we can the test looks cleaner.